### PR TITLE
ADMINUI-2296 VM - Cannot add/update machine tag 'triton.cns.disable'

### PR DESCRIPTION
**Pull request by Arc311**

### CODE REVIEW

**Comment by dan-luk** [comment]
🔢

**Comment by dan-luk** [comment]
🔢

**Comment by askfongjojo** [comment]
Hi @Arc311 - The UI doesn't surface any validation errors. Can we also make the change to pop up an error alert to the user when they have entered invalid values? e.g. triton.cns.disable=yes

**Comment by Arc311** [comment]
Please specify what other tags should be validated?

**Comment by dan-luk** [comment]
Hi @askfongjojo,

Denis means that historically there is no UI validation on the tag values except for allowed symbols,

- should we just add boolean validation for specific tag key of 'triton.cns.disable'?
- should we allow 'yes'/'no' to be treated like true/false?

**Comment by askfongjojo** [comment]
@dan-luk / @Arc311 - Got it. I was thinking of surfacing validation errors from vmapi but then realized that it was a workflow job and any error response would be async only. So please just add a specific boolean validation for 'triton.cns.disable' for true/false (AdminUI would just reflect what the API does so i think it's okay to not support yes/no).

File: www/js/components/pages/vm/index.jsx [hunk 1]
```diff
@@ -543,6 +544,15 @@ var VMPage = React.createClass({
         this.setState({editing: false});
     },
     _handleTagsSave: function (data) {
+        var cnsPropertyName = 'triton.cns.disable';
```
Comment (dan-luk) [inline www/js/components/pages/vm/index.jsx]: TRITON_CNS_DISABLE_TAG_KEY

File: www/js/components/pages/vm/index.jsx [hunk 2]
```diff
@@ -543,6 +544,15 @@ var VMPage = React.createClass({
...
+        if (Object.prototype.hasOwnProperty.call(data, cnsPropertyName) &&
+            !/^(true|false)$/.test(data[cnsPropertyName])) {
+            return adminui.vent.trigger('notification', {
+                level:'error',
+                message: 'Invalid value for \'' + cnsPropertyName + '\' tag.'
```
Comment (dan-luk) [inline www/js/components/pages/vm/index.jsx]: this is not helpful,

```Invalid value for \'' + cnsPropertyName + '\' tag.  Possible values are: true, false.'```

File: www/js/components/pages/vm/index.jsx [hunk 3]
```diff
@@ -544,12 +544,12 @@ var VMPage = React.createClass({
         this.setState({editing: false});
     },
     _handleTagsSave: function (data) {
-        var cnsPropertyName = 'triton.cns.disable';
-        if (Object.prototype.hasOwnProperty.call(data, cnsPropertyName) &&
-            !/^(true|false)$/.test(data[cnsPropertyName])) {
+        var TRITON_CNS_DISABLE_TAG_KEY = 'triton.cns.disable';
+        if (Object.prototype.hasOwnProperty.call(data, TRITON_CNS_DISABLE_TAG_KEY) &&
```
Comment (dan-luk) [inline www/js/components/pages/vm/index.jsx]:
```
var tagValue = data && data[TRITON_CNS_DISABLE_TAG_KEY];
if (tagValue !== ‘true’ && tagValue !== ‘false’) {
```

Comment (dan-luk) [inline www/js/components/pages/vm/index.jsx]: alternatively, you can put the code below into line 565 instead:

```
if (key === 'triton.cns.disable' && data[key] !== Boolean(data[key])) {
    return adminui.vent.trigger('notification', {
                  level:'error',
                 message: 'Invalid value for \'' + TRITON_CNS_DISABLE_TAG_KEY + '\' tag. Possible values are: true, false.'
              });
    }
```

Comment (Arc311) [inline www/js/components/pages/vm/index.jsx]: don't agree with second part of evaluation. Not empty string will always be true and strict comparison of different types will always be false. 
In other words 'value !== Boolean(value)' will always be true.

Comment (dan-luk) [inline www/js/components/pages/vm/index.jsx]: that's not true :)

I mean the values are already boolean at that point

File: lib/vms.js [hunk 4]
```diff
@@ -299,20 +299,22 @@ function updateTags(req, res, next) {
...
         if (err) {
             req.log.fatal(err);
             return next(err);
         }
         res.send(obj);
         return next();
-    });
+    }
+    if (!params.metadata) {
+        return req.sdc[req.dc].vmapi.deleteAllMetadata(
```
Comment (askfongjojo) [inline lib/vms.js]: This is not doing what is intended as the tag value is an empty json, not string. I deleted all tags and deleteAllMetadata didn't happen. Also setMetadata should be triggered in the else-condition. It got invoked and returned 409 for the empty json.

**Comment by dan-luk** [comment]
🔢

**Comment by askfongjojo** [comment]
merged as 848f710

**Pull request closed without merging**